File: template/.claude/hooks/post_edit.py
#!/usr/bin/env python3
"""Post-Edit hook: auto-formats Python files after Write/Edit.

Actions:
- Runs `ruff format` on Python files
- Shows lint warnings (non-blocking)

Requirements:
- ruff must be installed (pip install ruff)
"""

from __future__ import annotations

import os
import subprocess  # nosec: B404
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from utils import get_tool_input, log_hook_error, post_continue, read_hook_input

# Ruff configuration
RUFF_TIMEOUT_SECONDS = 10  # Max time for format/lint operations
RUFF_LINT_RULES = "E,W,F"  # Error, Warning, Pyflakes (fast, high-value checks)
MAX_LINT_WARNINGS = 5  # Limit output to avoid flooding Claude's context

# Claude Code tools that write files
# - Write: create new file
# - Edit: modify existing file (old_string → new_string)
# - MultiEdit: batch edits (reserved for future Claude Code feature)
FILE_WRITE_TOOLS = ("Write", "Edit", "MultiEdit")


def format_python_file(file_path: str) -> tuple[bool, str]:
    """Format Python file with ruff.

    Returns:
        (success, message)
    """
    try:
        result = subprocess.run(  # nosec: B603, B607
            ["ruff", "format", file_path],
            capture_output=True,
            text=True,
            timeout=RUFF_TIMEOUT_SECONDS,
        )
        if result.returncode == 0:
            return True, "formatted"
        return False, result.stderr or "format failed"
    except FileNotFoundError:
        return False, "ruff not found"
    except subprocess.TimeoutExpired:
        return False, "format timeout"
    except Exception as e:
        return False, str(e)


def check_lint_warnings(file_path: str) -> list[str]:
    """Check for lint warnings (non-blocking).

    Returns:
        List of warning messages
    """
    try:
        result = subprocess.run(  # nosec: B603, B607
            ["ruff", "check", file_path, f"--select={RUFF_LINT_RULES}"],
            capture_output=True,
            text=True,
            timeout=RUFF_TIMEOUT_SECONDS,
        )
        if result.stdout:
            return result.stdout.strip().split("\n")[:MAX_LINT_WARNINGS]
        return []
    except Exception:  # nosec B110 — fail-safe: lint error = skip linting
        return []


def main() -> None:
    try:
        data = read_hook_input()
        tool_name = data.get("tool_name", "")

        # Only process file-writing tools
        if tool_name not in FILE_WRITE_TOOLS:
            post_continue()
            return

        file_path = get_tool_input(data, "file_path") or ""

        # Only process Python files
        if not file_path.endswith(".py"):
            post_continue()
            return

        # Skip if file doesn't exist
        if not os.path.exists(file_path):
            post_continue()
            return

        messages = []

        # Format the file
        success, msg = format_python_file(file_path)
        if success:
            messages.append(f"ruff format: {os.path.basename(file_path)}")

        # Check for lint warnings (non-blocking)
        warnings = check_lint_warnings(file_path)
        if warnings:
            messages.append(f"lint warnings ({len(warnings)}):")
            messages.extend([f"  {w}" for w in warnings])

        if messages:
            post_continue("\n".join(messages))
        else:
            post_continue()

    except Exception as e:
        log_hook_error("post_edit", e)
        post_continue()  # Fail-safe: don't block on errors


if __name__ == "__main__":
    main()

File: template/.claude/hooks/post-edit.mjs
/**
 * Post-Edit hook: auto-formats Python files after Write/Edit.
 *
 * Actions:
 * - Runs `ruff format` on Python files
 * - Shows lint warnings (non-blocking)
 *
 * Requirements:
 * - ruff must be installed (pip install ruff)
 */

import { existsSync } from 'fs';
import { basename, resolve } from 'path';
import { execFileSync } from 'child_process';
import { debugLog, debugTiming, getProjectDir, getToolInput, logHookError, postContinue, readHookInput } from './utils.mjs';

// Ruff configuration
const RUFF_TIMEOUT_MS = 10000;
const RUFF_LINT_RULES = 'E,W,F';
const MAX_LINT_WARNINGS = 5;

// Claude Code tools that write files
const FILE_WRITE_TOOLS = ['Write', 'Edit', 'MultiEdit'];

function formatPythonFile(filePath) {
  try {
    execFileSync('ruff', ['format', filePath], { timeout: RUFF_TIMEOUT_MS, stdio: 'pipe' });
    return { success: true, msg: 'formatted' };
  } catch (e) {
    if (e.code === 'ENOENT') return { success: false, msg: 'ruff not found' };
    if (e.killed) return { success: false, msg: 'format timeout' };
    return { success: false, msg: String(e.message || e) };
  }
}

function checkLintWarnings(filePath) {
  try {
    const result = execFileSync('ruff', ['check', filePath, `--select=${RUFF_LINT_RULES}`], {
      timeout: RUFF_TIMEOUT_MS,
      stdio: 'pipe',
      encoding: 'utf-8',
    });
    return result ? result.trim().split('\n').slice(0, MAX_LINT_WARNINGS) : [];
  } catch (e) {
    // ruff check returns non-zero when warnings found — output is in stderr/stdout
    const output = e.stdout || '';
    return output ? output.trim().split('\n').slice(0, MAX_LINT_WARNINGS) : [];
  }
}

function main() {
  const timer = debugTiming('post-edit');
  try {
    const data = readHookInput();
    const toolName = data.tool_name || '';

    if (!FILE_WRITE_TOOLS.includes(toolName)) {
      debugLog('post-edit', 'skip', { reason: 'not_write_tool', tool: toolName });
      timer.end('skip');
      postContinue();
      return;
    }

    const filePath = getToolInput(data, 'file_path') || '';
    debugLog('post-edit', 'input', { tool: toolName, file: filePath });

    if (!filePath.endsWith('.py')) {
      debugLog('post-edit', 'skip', { reason: 'not_python' });
      timer.end('skip');
      postContinue();
      return;
    }

    if (!existsSync(filePath)) {
      debugLog('post-edit', 'skip', { reason: 'file_not_found' });
      timer.end('skip');
      postContinue();
      return;
    }

    // Project boundary check (F-018)
    const projectDir = getProjectDir();
    const absPath = resolve(filePath);
    if (!absPath.startsWith(projectDir)) {
      debugLog('post-edit', 'skip', { reason: 'outside_project' });
      timer.end('skip');
      postContinue();
      return;
    }

    const messages = [];

    const { success } = formatPythonFile(filePath);
    if (success) {
      messages.push(`ruff format: ${basename(filePath)}`);
    }

    const warnings = checkLintWarnings(filePath);
    if (warnings.length > 0) {
      messages.push(`lint warnings (${warnings.length}):`);
      messages.push(...warnings.map(w => `  ${w}`));
    }

    debugLog('post-edit', 'continue', { formatted: success });
    timer.end('continue');

    if (messages.length > 0) {
      postContinue(messages.join('\n'));
    } else {
      postContinue();
    }
  } catch (e) {
    debugLog('post-edit', 'error', { error: String(e) });
    timer.end('error');
    logHookError('post_edit', e);
    postContinue(); // Fail-safe: don't block on errors
  }
}

main();
